dopamine/discrete_domains/run_experiment.py

```diff
@@ -600,18 +609,57 @@ def _run_iterations(self):
     time an iteration completes a new one starts until the right number of
     iterations is run.
     """
-    # TODO(aarg): Change the thread management to an implementation with queues.
-    threads = []
+    # TODO(aarg): Change the thread management to an implementation with queues
+    # and with a fix number of thread workers. With the revamp, training might
+    # go inline to this method.
+    training_worker = threading.Thread(target=self._run_training_steps)
+    training_worker.start()
+    threads = [training_worker]
     for iteration in range(self._start_iteration, self._num_iterations):
-      self._running_iterations.acquire()
+      self._running_iterations.put(1)
       thread = threading.Thread(
           target=self._run_one_iteration, args=(iteration,))
       thread.start()
       threads.append(thread)
-    # Wait for all running iterations to complete.
+
+    # Wait for all tasks to complete.
+    self._running_iterations.join()
+    self._remaining_training_steps.join()
+    # Indicate training step thread to stop.
+    self._remaining_training_steps.put(None)
+    # Wait for all running threads to complete.
     for thread in threads:
       thread.join()
 
+  def _begin_episode(self, observation):
+    # Increments training steps and blocks if training is too slow.
```

Comment:
Is there a goal behind blocking, or is it just a necessary consequence of having only one `training_worker`?  I'm a little confused about the goal behind these overrides of `Runner._begin_episode`.

Reply:
Regarding blocking:
in the case where experience generation is very fast compared to training step, you can end up with the situation where almost all your experience generation have completed and almost none of the training steps have run (extrapolating) which is an issue. So you want to avoid training to accumulate delay compared to experience generation and ensure that training occurs at the same pace as experience generation.
For this purpose the training steps queue has a limit which is the number of workers. If it's full we block experience generation until more training steps have been completed.

This is not something we would face in kspice where in anycase, training is much faster.


Regarding overriding `_begin_episode` the goal is simply to allow to increment the number of training steps to run each time we run an environment step.

Reply:
Aha!  Thanks -- that pulls things together for me.

I think it might be worth writing up an overall strategy into 2-3 paragraphs for a README.  That README could both be eventual documentation, and also a way to help us look at the overall design and not miss the forest for the implementation trees.

Aaand, this is easy for me to say because it's asking for you to do work.

Reply:
Done.